### PR TITLE
Create a MOM_diag_manager_infra module

### PR DESCRIPTION
**Pull request by MJHarrison-GFDL**
references and documentation for shared diagnostic infrastructure

### CODE REVIEW

File: src/framework/MOM_diag_manager_infra.F90 [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+!> Returns true if the argument data are successfully passed to a diagnostic manager
+!!  with the indicated unique reference id, false otherwise.
+logical function send_data_2d_r8(diag_field_id, field, time, is_in, js_in, mask, rmask, &
+                            &  ie_in, je_in, weight, err_msg)
+    integer, intent(in) :: diag_field_id !< A unique identifier for this data to the diagnostic manager
+    real(kind=8), dimension(:,:), intent(in) :: field !< A rank 1 array of floating point values being recorded
+    type (time_type), intent(in), optional :: time  !< The time for the current record.
+    logical, intent(in), dimension(:,:), optional :: mask !< An optional rank 1 logical mask.
+    real, intent(in), dimension(:,:), optional :: rmask !< An optional rank 1 mask array
+    integer, intent(in), optional :: is_in, ie_in !< An optional range for subsetting the data being recorded.
```
Comment (adcroft) [inline src/framework/MOM_diag_manager_infra.F90:208]: Two variables but one comment is the source of your doxygen errors

**Comment by marshallward** [comment]
https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/11987

**Comment by marshallward** [comment]
The PGI compiler produced errors, shown below (thanks to @adcroft for pointing them out).

This could be related to the bug that @Hallberg-NOAA encountered, but perhaps you just need to extend these changes to the ice shelf diag mediator.
```
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 294)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 294)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 299)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 299)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 304)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 304)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 307)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 307)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 312)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 312)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 322)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 322)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 329)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 329)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 333)
build/pgi/repro/dynamic/mom6/libmom6.a: PGF90-S-0155-Ambiguous interfaces for generic procedure send_data (../../../../../..//src/ice_shelf/MOM_ice_shelf_diag_mediator.F90: 333)
```

**Review by Hallberg-NOAA — COMMENTED** [review]
Thank you for this PR, it is very close to where I think it should be. The changes to the interface names in MOM_ice_shelf_diag_mediator.F90 avoid reusing the names from MOM_diag_mediator.F90 are particularly helpful, and all of the interfaces in the new MOM_diag_manager_init.F90 are now nicely documented.

There are three minor issues with this commit that would like to raise (and then perhaps just dismiss). The first is that, in the end, this series of commits turned into an effective renaming of MOM_diag_manager.F90 to MOM_diag_manager_infra.F90, which is fine, but the way that the commits played out it made the Git history look like the creation of one new file and the deletion of the old one, so the file history is effectively lost and blame will not work. More than half the lines in this file were newly added by this series of commits, and most of the other lines were recent additions, so perhaps this is not problematic, but if there were a convenient way to reconnect the two files' history it would be nice.

  The second issue is that MOM_obsolete_diagnostics.F90 should not be accessing anything in MOM_diag_manager_infra.F90 directly. This might mean that we need to add a new thin layer to MOM_diag_mediator.F90, say register_field_unlogged(), as a wrapper to register_static_field_fms() from MOM_diag_manager_infra.F90, and then use this interface in MOM_obsolete_diagnostics.F90. Alternately, maybe the answer is just to move the function found_in_diagtable() from MOM_obsolete_diagnostics.F90 to MOM_diag_mediator.F90.

  The third, and least substantial, issue is in the naming of some of the routines from the MOM_diag_manager_infra as ..._fms_wrapper.  At some point, these might not be wrapping FMS calls, but something else, and we want to have names that describe their generic functionality so that we won't have to these names again if we could avoid it. Would a name like send_data_infra() work better than send_data_fms_wrapper()?

  Reconnecting the file history for MOM_diag_manager.F90 is the only one of these concerns that could not be addressed by accepting this PR, which mostly achieves its objectives, and then revising to address the other minor points.

**Comment by MJHarrison-GFDL** [comment]
1. I anticipate @marshallward will squash these commits. 
2. option 2 taken
3. ..._fms_wrapper functions are currently pass through interfaces to FMS routines with more or less an identical list of arguments, so I thought it less confusing to call it a fms_wrapper.  I concur with the idea that renaming to ..._infra makes sense only if we agree on the current interface are suitable for generic infrastructure calls.

**Review by Hallberg-NOAA — APPROVED** [review]
These changes provide a useful set of explicit interfaces to all of the FMS diag_manager routines that are called by MOM6.  This PR has passed all of the TC testing and the pipeline testing at https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/12007.

**Pull request merged**